engine/grid.py

```diff
@@ -314,59 +318,109 @@ def find_border_nodes(self):
                     if leftmost_node_pos is None or col < leftmost_node_pos[1]:
                         leftmost_node = node
                         leftmost_node_pos = (row, col)
+                if node.is_active and self.is_on_border(row, col, rows , cols):
+                    # check if this is an active node and on the border
+                    self.nodes[row][col].is_border = True
+                    if rightmost_node_pos is None or row > rightmost_node_pos[0]:
+                        rightmost_node = node
+                        rightmost_node_pos = (row, col)
+                    
+                    
+                        
         self.border_nodes = border_list
         self.leftmost_node = leftmost_node
         self.leftmost_node_pos = leftmost_node_pos
+        self.rightmost_node = rightmost_node
+        self.rightmost_node_pos = rightmost_node_pos
 
-    # Return bottom most node that is activated in the right column
-
+    
+    ##Return bottom most node that is activated in the right column
     def bottom_rightmost_node(self, pos):
+        """
+        A helper function for get_all_lawnmower_waypoints_adjustable, finds the rightmost
+        node of the row above current pos.
+
+        Parameters:
+        parameter1: pos
+        precondition: None
 
+        returns:  Returns the furthest right node on the row above the current pos, 
+        or return None if there is no nodes in the row above.
+        """
         # node_info: (node, row, col)
         candidate_nodes = [node_info for node_info in self.border_nodes if node_info[2] == pos[1]+1]
         if (candidate_nodes == []):
             return None
         else:
             node_info = min(candidate_nodes,key=lambda node_info: node_info[1])
             return (node_info[1],node_info[2])
+    
+    def bottom_leftmost_node(self, pos):
+        """
+        A helper function for get_all_lawnmower_waypoints_adjustable, finds the leftmost
+        node of the row above current pos.
 
+        Parameters:
+        parameter1: pos
+        precondition: None
+
+        returns:  Returns the furthest left node on the row above the current pos, 
+        or return None if there is no nodes in the row above.
+        """
+        # node_info: (node, row, col)
+        candidate_nodes = [node_info for node_info in self.border_nodes if node_info[2] == pos[1]+1]
+        if (candidate_nodes == []):
+            return None
+        else:
+            node_info = max(candidate_nodes,key=lambda node_info: node_info[1])
```

Comment:
It seems like this is the only change between bottommost left and right. What do you think about combining those two functions into one general function such as
`next_row_starting_node(self, pos, current_direction)`

Rather than bottommost left/right looking at ur spec and the code again, the code gives you the next node in the row above you want to traverse to once you reach an edge. We could take Direction(Enum) outside of get all lawnmower and send it in as a parameter to determine whether you should min/max

Reply:
This also made me realize that we should eventually expand between rowbase traversal and colummbase traversal since rn our current lawnmower impl used in sim_trajectory is column based

If we wanna name the above, we could likewise call the column-base equivalent next_column_starting_node...